src/rpg_database/character_sheet/FocusesSet.java

```diff
@@ -0,0 +1,171 @@
+package rpg_database.character_sheet;
+
+import java.util.ArrayList;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+import rpg_database.character_sheet.exceptions.InvalidFocusesSetException;
+import rpg_database.character_sheet.interfaces.CustomSetter;
+
+public class FocusesSet implements Set<Focus>, CustomSetter<FocusesSet> {
+
+	private HashSet<Focus> focusesSet;
+
+	public FocusesSet() {
+		focusesSet = new HashSet<Focus>();
+	}
+
+	public FocusesSet(Focuses... focuses) {
+		focusesSet = new HashSet<Focus>();
+		focusesSet = findDuplicates(focuses);
+
+	}
+
+	@Override
+	public Class<FocusesSet> getImplementingClass() {
+		return FocusesSet.class;
+	}
+
+	@Override
+	public void setSelfInSheet(CharacterSheet characterSheet) {
+		characterSheet.characterData.put(Fields.FOCUSES, this);
+	}
+
+	@Override
+	public int size() {
+		return focusesSet.size();
+	}
+
+	@Override
+	public boolean isEmpty() {
+		return focusesSet.isEmpty();
+	}
+
+	@Override
+	public boolean contains(Object o) {
+		return focusesSet.contains(o);
+	}
+
+	@Override
+	public Iterator<Focus> iterator() {
+		return focusesSet.iterator();
+	}
+
+	@Override
+	public Object[] toArray() {
+		return focusesSet.toArray();
+	}
+
+	@Override
+	public <T> T[] toArray(T[] a) {
+		return focusesSet.toArray(a);
+	}
+
+	/**
+	 * Does some thing in old style.
+	 *
+	 * @deprecated use {@link #new()} instead.
+	 */
+	@Deprecated
+	@Override
+	public boolean add(Focus focus) {
+		return focusesSet.add(focus);
+	}
+
+	public boolean add(Focuses focus) {
+		Focus focusInDb = getFocus(focus, focusesSet);
+		if (focusInDb == null) {
+			Focus tempFocus = new Focus(focus);
+			focusesSet.add(tempFocus);
+			return tempFocus.isFocusImproved();
+		} else {
+			if (focusInDb.isFocusImproved()) {
+				throw new InvalidFocusesSetException(focus.name() + " is already improved!");
+			}
+			focusInDb.makeFocusImproved();
+			return focusInDb.isFocusImproved();

```

Comment:
Itt is miért az isFocusImproved-ot adod vissza az add logikus kimeneteként? Az add végeredménye, hogy improved -e a focus? Ezt átgondoltad?

Reply:
Mert csak akkor lesz improved ha sikerült felvenni a focust, ami egyfajta kiterjesztése az add metódusnak értelmezésem szerint. Nem a legszerencsésebb, de jobbnak tűnt mint a return true;

Reply:
Hát azért nem szerencsés ez, mert add esetén a false azt jelenti hogy már benne van a setben, ergo nem adjuk ismét hozzá. Nálunk ugye ez kicsit komplikáltabb, mert ha nem lehet már hozzáadni akkor dobunk, de ugye hozzá lehet akkor is adni amikor jelen van már az elem ha improveolod... Én azt mondanám hogy inkább dobj vissza true-t, mert az isFocusImproved kimenete igen összezavaró lehet, mi van ha valaki használja feltételként figyelmetlenségből...Én itt arra szavaznék hogy értelmesebb tartani az eredeti logikát a kimenetre amennyire lehet, de ha meg tudsz győzni hogy mit nyerünk ezen, akkor nyitott vagyok rá hogy maradjon. Csak nem látom hogy mi extrát kapunk belőle a true visszaadásával szemben.

Reply:
Hát azért nem szerencsés ez, mert add esetén a false azt jelenti hogy már benne van a setben, ergo nem adjuk ismét hozzá. Nálunk ugye ez kicsit komplikáltabb, mert ha nem lehet már hozzáadni akkor dobunk, de ugye hozzá lehet akkor is adni amikor jelen van már az elem ha improveolod... Én azt mondanám hogy inkább dobj vissza true-t, mert az isFocusImproved kimenete igen összezavaró lehet, mi van ha valaki használja feltételként figyelmetlenségből...Én itt arra szavaznék hogy értelmesebb tartani az eredeti logikát a kimenetre amennyire lehet, de ha meg tudsz győzni hogy mit nyerünk ezen, akkor nyitott vagyok rá hogy maradjon. Csak nem látom hogy mi extrát kapunk belőle a true visszaadásával szemben.